postgres/pgdump.go

```diff
@@ -684,57 +689,62 @@ func toForeignKeys(fk constraint) (fkey schema.ForeignKey) {
 }
 
 // getCols extracts and returns the column names for an InsertStatement.
-func getCols(conv *internal.Conv, table string, l []nodes.Node) (cols []string, err error) {
-	for _, n := range l {
-		switch r := n.(type) {
-		case nodes.ResTarget:
-			if r.Name != nil {
-				cols = append(cols, *r.Name)
+func getCols(conv *internal.Conv, table string, nodes []*pg_query.Node) (cols []string, err error) {
+	for _, n := range nodes {
+		switch r := n.GetNode().(type) {
+		case *pg_query.Node_ResTarget:
+			if r.ResTarget.Name != "" {
+				cols = append(cols, r.ResTarget.Name)
 			}
 		default:
-			return nil, fmt.Errorf("expecting ResTarget node but got %v node while processing Cols", reflect.TypeOf(r))
+			return nil, fmt.Errorf("expecting ResTarget node but got %v node while processing Cols", printNodeType(r))
 		}
 	}
 	return cols, nil
 }
 
 // getVals extracts and returns the values for an InsertStatement.
-func getVals(conv *internal.Conv, l [][]nodes.Node, n nodes.InsertStmt) (values []string) {
-	for _, vl := range l {
-		for _, v := range vl {
-			switch c := v.(type) {
-			case nodes.A_Const:
-				switch st := c.Val.(type) {
-				case nodes.String:
-					values = append(values, st.Str)
-				case nodes.Integer:
-					// For uniformity, convert to string and handle everything in
-					// dataConversion(). If performance of insert statements becomes a
-					// high priority (it isn't right now), then consider preserving int64
-					// here to avoid the int64 -> string -> int64 conversions.
-					values = append(values, strconv.FormatInt(st.Ival, 10))
+func getVals(conv *internal.Conv, vll []*pg_query.Node, n *pg_query.InsertStmt) (values []string) {
```

Comment:
I'm pretty sure the original code here is just broken (and the new code preserves the bug).

In short, getVals should actually return [][]string.

Ok, the long details ...

In pgdump, the default way to write out data to the dump file is to use COPY-FROM. However, you can also get pgdump to generate INSERT statements. The very first version of harbourbridge used the insert version (it was kind of easier to start with). Now, when pgdump generates INSERTS, it only generates them with one row at a time i.e.

INSERT INTO public.products (product_id, description, price, date_added) VALUES ('abc-123', 'Blue suede shoes', '141.99', '2020-06-06');

And that was all that I implemented at the time.
However, PostgreSQL supports multi-row inserts i.e.:

INSERT INTO public.products (product_id, description, price, date_added) VALUES
('abc-123', 'Blue suede shoes', '141.99', '2020-06-06'),
('axd-673', 'Antique typewriter', '99.99', '2020-06-07'),
('zxi-631', 'Glass vase', '55.50', '2020-06-10');

These get completely mangled by getVals! You get something like:

vals = ["abc-123" "Blue suede shoes" "141.99" "2020-06-06" "axd-673" "Antique typewriter" "99.99" "2020-06-07" "zxi-631" "Glass vase" "55.50" "2020-06-10"]

i.e. all of the values from all of the rows concatenated together!

To fix this, we should:
a) return [][]string from getVals
b) change copyOrInsert to handle [][]string for vals (maybe we should call it rows?)
c) change ProcessPgDump to handle multi-row inserts i.e. iterate over the [][]string and call ProcessDataRow on each row.
d) add a test for multi-row INSERT statements (in fact, this should probably be the first step!)

I suggest we put a TODO for this in the code and defer for another PR. I'll file an issue for this to capture the state.

Reply:
I've added issue #176 for this.

Reply:
Thanks for the detailed history, I was also kind of wondering about the same when I looked at this code but didn't change functionality in this PR. Doing this in a follow up with appropriate testing makes sense to me and I have assigned the github issue to myself. Added TODO. Done @https://github.com/cloudspannerecosystem/harbourbridge/pull/172/commits/3044120ca85a3b4386f12ca5a583fe4436cd0c54.